File: iminling-core/src/main/java/com/iminling/core/config/value/GlobalReturnValueHandler.java
package com.iminling.core.config.value;

import com.iminling.common.json.JsonUtil;
import com.iminling.core.annotation.EnableResolve;
import com.iminling.model.common.MessageCode;
import com.iminling.model.common.ResultModel;
import java.lang.reflect.AnnotatedElement;
import java.lang.reflect.Method;
import lombok.extern.slf4j.Slf4j;
import org.springframework.core.MethodParameter;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.http.converter.HttpMessageConverter;
import org.springframework.http.server.ServerHttpRequest;
import org.springframework.http.server.ServerHttpResponse;
import org.springframework.web.bind.annotation.RestControllerAdvice;
import org.springframework.web.servlet.mvc.method.annotation.ResponseBodyAdvice;
import org.springframework.web.servlet.mvc.method.annotation.SseEmitter;
import org.springframework.web.servlet.mvc.method.annotation.StreamingResponseBody;

@Slf4j
@RestControllerAdvice
public class GlobalReturnValueHandler implements ResponseBodyAdvice<Object> {

    @Override
    public boolean supports(MethodParameter returnType, Class<? extends HttpMessageConverter<?>> converterType) {
        Class<?> parameterType = returnType.getParameterType();
        if (ResultModel.class.isAssignableFrom(parameterType)
                || ResponseEntity.class.isAssignableFrom(parameterType)
                || StreamingResponseBody.class.isAssignableFrom(parameterType)
                || StreamingResponseBody.class.isAssignableFrom(parameterType)
                || SseEmitter.class.isAssignableFrom(parameterType)) {
            return false;
        }
        Method method = returnType.getMethod();
        AnnotatedElement annotatedElement = returnType.getAnnotatedElement();
        EnableResolve enableResolve = annotatedElement.getAnnotation(EnableResolve.class);
        if (enableResolve == null) {
            if (method != null) enableResolve = method.getAnnotation(EnableResolve.class);
        }
        if (enableResolve == null) {
            enableResolve =  returnType.getContainingClass().getAnnotation(EnableResolve.class);
        }
        if (enableResolve == null) return false;
        EnableResolve.ResolveStrategy value = enableResolve.value();
        return value.equals(EnableResolve.ResolveStrategy.ALL) || value.equals(EnableResolve.ResolveStrategy.RETURN_VALUE);
    }

    @Override
    public Object beforeBodyWrite(Object body, MethodParameter returnType, MediaType selectedContentType,
                                  Class<? extends HttpMessageConverter<?>> selectedConverterType, ServerHttpRequest request, ServerHttpResponse response) {
        ResultModel resultModel = new ResultModel(MessageCode.RESULT_OK);
        resultModel.setData(body);
        boolean isString = returnType.getParameterType() == String.class;
        if (isString) {
            String result = JsonUtil.obj2Str(resultModel);
            response.getHeaders().setContentType(MediaType.APPLICATION_JSON);
            return result;
        }
        return resultModel;
    }

    private boolean isPrimitiveOrVoid(Class<?> returnType) {
        return returnType.isPrimitive()
                || Number.class.isAssignableFrom(returnType)
                || CharSequence.class.isAssignableFrom(returnType)
                || Character.class.isAssignableFrom(returnType)
                || Boolean.class.isAssignableFrom(returnType);
    }

}

File: iminling-core/src/main/kotlin/com/iminling/core/config/value/GlobalReturnValueHandler.kt
package com.iminling.core.config.value

import com.iminling.common.json.JsonUtil
import com.iminling.core.annotation.EnableResolve
import com.iminling.core.annotation.EnableResolve.ResolveStrategy
import com.iminling.core.constant.MessageCode
import org.springframework.core.MethodParameter
import org.springframework.http.MediaType
import org.springframework.http.ResponseEntity
import org.springframework.http.converter.HttpMessageConverter
import org.springframework.http.server.ServerHttpRequest
import org.springframework.http.server.ServerHttpResponse
import org.springframework.web.bind.annotation.RestControllerAdvice
import org.springframework.web.servlet.mvc.method.annotation.ResponseBodyAdvice
import org.springframework.web.servlet.mvc.method.annotation.SseEmitter
import org.springframework.web.servlet.mvc.method.annotation.StreamingResponseBody

/**
 * @author  yslao@outlook.com
 * @since  2021/11/26
 */
@RestControllerAdvice
class GlobalReturnValueHandler: ResponseBodyAdvice<Any> {
    override fun supports(returnType: MethodParameter, converterType: Class<out HttpMessageConverter<*>>): Boolean {
        val parameterType = returnType.parameterType
        if (ResultModel::class.java.isAssignableFrom(parameterType)
            || ResponseEntity::class.java.isAssignableFrom(parameterType)
            || StreamingResponseBody::class.java.isAssignableFrom(parameterType)
            || StreamingResponseBody::class.java.isAssignableFrom(parameterType)
            || SseEmitter::class.java.isAssignableFrom(parameterType)
        ) {
            return false
        }
        val method = returnType.method
        val annotatedElement = returnType.annotatedElement
        var enableResolve = annotatedElement.getAnnotation(EnableResolve::class.java)
        if (enableResolve == null) {
            if (method != null) enableResolve = method.getAnnotation(EnableResolve::class.java)
        }
        if (enableResolve == null) {
            enableResolve = returnType.containingClass.getAnnotation(EnableResolve::class.java)
        }
        if (enableResolve == null) return false
        val value = enableResolve.value
        return value == ResolveStrategy.ALL || value == ResolveStrategy.RETURN_VALUE
    }

    override fun beforeBodyWrite(
        body: Any?,
        returnType: MethodParameter,
        selectedContentType: MediaType,
        selectedConverterType: Class<out HttpMessageConverter<*>>,
        request: ServerHttpRequest,
        response: ServerHttpResponse
    ): Any {
        val resultModel = ResultModel<Any>(MessageCode.RESULT_OK)
        resultModel.data = body
        val isString = returnType.parameterType == String::class.java
        if (isString) {
            val result = JsonUtil.obj2Str(resultModel)
            response.headers.contentType = MediaType.APPLICATION_JSON
            return result
        }
        return resultModel
    }

    private fun isPrimitiveOrVoid(returnType: Class<*>): Boolean {
        return (returnType.isPrimitive
                || Number::class.java.isAssignableFrom(returnType)
                || CharSequence::class.java.isAssignableFrom(returnType)
                || Char::class.java.isAssignableFrom(returnType)
                || Boolean::class.java.isAssignableFrom(returnType))
    }
}